src/main/java/org/broadinstitute/hellbender/tools/spark/utils/HopscotchHashSet.java

```diff
@@ -0,0 +1,542 @@
+        kryo.setReferences(false);
+        capacity = input.readInt();
+        size = 0;
+        buckets = (T[])new Object[capacity];
+        status = new byte[capacity];
+        int size = input.readInt();
+        while ( size-- > 0 ) {
+            add((T)kryo.readClassAndObject(input));
+        }
+        kryo.setReferences(oldReferencesState);
+    }
+
+    private void serialize( final Kryo kryo, final Output output ) {
+        final boolean oldReferencesState = kryo.getReferences();
+        // this is actually screwy -- what we need to do is to turn off just the top-level reference resolution as
+        // if we were using the ObjectOutputStream's writeUnshared method.
+        // we're a set, after all, and each object will be unique.  but we don't know what's going on inside
+        // the objects in the set -- they might well have shared references to other objects.
+        // but with kryo there doesn't seem to be a way to do this -- you have to turn off all reference resolution.
+        kryo.setReferences(false);
+        output.writeInt(capacity);
+        output.writeInt(size);
+        int count = 0;
+        for ( int idx = 0; idx != capacity; ++idx ) {
+            if ( isChainHead(idx) ) {
+                kryo.writeClassAndObject(output, buckets[idx]);
+                count += 1;
+            }
+        }
+        for ( int idx = 0; idx != capacity; ++idx ) {
+            final T val = buckets[idx];
+            if ( val != null && !isChainHead(idx) ) {
+                kryo.writeClassAndObject(output, val);
+                count += 1;
+            }
+        }
+        kryo.setReferences(oldReferencesState);
+
+        if ( count != size ) {
+            throw new IllegalStateException("Failed to serialize the expected number of objects: expected="+size+" actual="+count+".");
+        }
+    }
+
+    @Override
+    public boolean contains( final Object value ) {
+        if ( value == null ) return false;
+        int bucketIndex = hashToIndex(value.hashCode());
+        if ( !isChainHead(bucketIndex) ) return false;
+        if ( buckets[bucketIndex].equals(value) ) return true;
+        int offset;
+        while ( (offset = getOffset(bucketIndex)) != 0 ) {
+            bucketIndex = getIndex(bucketIndex, offset);
+            if ( buckets[bucketIndex].equals(value) ) return true;
+        }
+        return false;
+    }
+
+    public int capacity() { return capacity; }
+
+    @Override
+    public int size() { return size; }
+
+    @Override
+    public Iterator<T> iterator() { return new CompleteIterator(); }
+
+    /**
+     * This special iterator type allows you to traverse individual hash buckets.
+     * This lets you implement a multi-map on this collection by using a <K,V>-ish element type having a hashCode that
+     * depends only on K.
+     */
+    public Iterator<T> bucketIterator( final int hashVal ) { return new BucketIterator(hashVal); }
```

Comment:
I think it would make things more clear if you made this method (or a sister method) take an argument of type `T` and compute the hash within this object.